pkg/backup/remap_crd_version_action.go

```diff
@@ -103,24 +110,51 @@ func (a *RemapCRDVersionAction) Execute(item runtime.Unstructured, backup *v1.Ba
 		if c.Type == apiextv1.NonStructuralSchema {
 			log.Debug("CRD is a non-structural schema")
 
-			if err := setV1beta1Version(item); err != nil {
+			item, err = fetchV1beta1CRD(crd.Name, a.betaCRDClient)
+			if err != nil {
 				return nil, nil, err
 			}
-
 			break
 		}
 	}
 
+	// CRD has the field "Spec.PreserveUnknownFields". This meant it was a v1beta1 CRD and cannot be restored as a v1 CRD - validation webhooks will fail.
+	// TODO: add a test case
+	if crd.Spec.PreserveUnknownFields {
```

Comment:
Not a must-do, but I think this whole plugin could be made a little clearer if we extract 3 helper functions, each of which have a signature of `func(apiextv1.CustomResourceDefinition) bool`, for each of the three conditions under which we want to back up the `v1beta1` representation:
- does not have a schema
- has the NonStructuralSchema condition
- has spec.preserverunknownfields = true

Then the main `Execute` function can just call each of those in turn to see if the provided CRD matches any of those conditions, and if so, backs it up as `v1beta1`.

I think this structure would be more readable.

Again, not blocking - if you want to do it in a different PR, or just put a `// TODO`, that's OK with me for now.

Reply:
Yeah, I can put functions in to name the cases, that makes sense. I can make that a TODO so we can get this in a different PR, but I think it should be in for the 1.4 release just to limit technical debt.